grpc-client-server/README.md

```diff
@@ -0,0 +1,188 @@
+
+## Implement prober logic
+
+Let's modify the prober service slightly. Instead of the simple HelloWorld greeting, we are going to implement a service that will probe a HTTP endpoint N times and return the average time to fetch that endpoint to the client. 
+
+This tool could be useful in a real operational context. Imagine that we want to verify that our site is available and has acceptable latency from many different locations around the world. We might run several instances of our prober server in
+different regions. 
+
+Change your prober request and response to look like this and regenerate your generated go code.
+```
+// The request message
+message ProbeRequest {
+  string endpoint = 1;    // Endpoint to probe
+  int32 repetitions = 2;  // How many probes to execute
+}
+
+// The response message 
+message ProbeReply {
+  float average = 1;    // Average response time (in milliseconds)
+}
+```
+
+Update your client to read the endpoint and number of repetitions from the [command line](https://gobyexample.com/command-line-arguments).
+Then update your server to execute the probe: do a HTTP fetch of `endpoint` the specified number of times.
+Use the standard [`net/http` package](https://pkg.go.dev/net/http).
+
+Time each fetch using the `time` package:
+
+```
+  	start := time.Now()
+	resp, _:= http.Get(endpoint)
+    elapsed := time.Since(start)
+```
+
+Add up all the elapsed times, divide by the number of repetitions, and return the average to the client.
+The client should print out the average value received.
+You can do arithmetic operations like addition and division on `time.Duration` values in Go.
+However, under the hood `time.Duration` is an `int64`. To divide by an `int32` and return a `float` will need to do some [type conversions](https://go.dev/tour/basics/13).
+
+## Add a client timeout
+
+Maybe the site you are probing is very slow, or the number of repetitions is very high.
+Either way, you don't want your program to wait forever. 
+
+On the client side, add a [timeout](https://pkg.go.dev/context#WithTimeout) to stop waiting after 1 second].
+
+Run your client against some website - how many repetitions do you need to see your client timeout?
+
+## Extra Challenges: Errors, Continuous Probing, Serve Prometheus Metrics
+
+These sections are optional - do them for an extra challenge if time permits.
+
+### Part 1:
+How do we know if the HTTP fetch succeeded at the server? Add a check to make sure it did.
+How should we deal with errors, i.e. if the endpoint isn't found?
+Modify your code and proto format to handle these cases.
+
+### Part 2: Run probes on a schedule
+Instead of getting probe times via the gRPC client we might consider running the endpoint probes on a schedule. Change the
+prober_server to accept a frequency argument instead of the old `repetitions` parameter, 
+and continue running probes every N seconds forever. 
```

Comment:
Why would we do this? Any good resources?

Change vs add an option? Goes back to managing change through time, and avoiding breaking changes. Can we teach the good habit? We don't have to be explicit about why, tho that would be nice.